src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,407 @@
+            );
+        }
+
+        if (start >= howMany || howMany == 0) {
+            return EMPTY_TOPDOCS;
+        }
+
+        int size = howMany - start;
+        ScoreDoc[] results = new ScoreDoc[size];
+
+        // Get the requested results from pq.
+        populateResults(results, size, pq);
+
+        return new TopFieldDocs(new TotalHits(totalHits, totalHitsRelation), results, sortFields);
+    }
+
+    /*
+      Results are converted in the FieldDocs and the value of the field on which the sorting is applied has been added in the FieldDoc.
+     */
+    private void populateResults(ScoreDoc[] results, int howMany, PriorityQueue<FieldValueHitQueue.Entry> pq) {
+        FieldValueHitQueue<FieldValueHitQueue.Entry> queue = (FieldValueHitQueue<FieldValueHitQueue.Entry>) pq;
+        for (int i = howMany - 1; i >= 0 && pq.size() > 0; i--) {
+            // adding to array if index is within [0..array_length - 1]
+            if (i < results.length) {
+                FieldValueHitQueue.Entry entry = queue.pop();
+                final int n = queue.getComparators().length;
+                final Object[] fields = new Object[n];
+                for (int j = 0; j < n; ++j) {
+                    fields[j] = queue.getComparators()[j].value(entry.slot);
+                }
+
+                results[i] = new FieldDoc(entry.doc, entry.score, fields);
+            }
+        }
+    }
+
+    // Add the entry in the Priority queue
+    private void add(int slot, int doc, FieldValueHitQueue<FieldValueHitQueue.Entry> compoundScore, int subQueryNumber, float score) {
+        FieldValueHitQueue.Entry bottomEntry = new FieldValueHitQueue.Entry(slot, docBase + doc);
+        bottomEntry.score = score;
+        bottom = compoundScore.add(bottomEntry);
+        // The queue is full either when totalHits == numHits (in SimpleFieldCollector), in which case
+        // slot = totalHits - 1, or when hitsCollected == numHits (in PagingFieldCollector this is hits
+        // on the current page) and slot = hitsCollected - 1.
+        assert slot < numHits;
+        queueFull[subQueryNumber] = slot == numHits - 1;
+    }
+
+    private void updateBottom(int doc, FieldValueHitQueue<FieldValueHitQueue.Entry> compoundScore) {
+        bottom.doc = docBase + doc;
+        bottom = compoundScore.updateTop();
+    }
+
+    private boolean canEarlyTerminate(Sort searchSort, Sort indexSort) {
+        return canEarlyTerminateOnDocId(searchSort) || canEarlyTerminateOnPrefix(searchSort, indexSort);
+    }
+
+    private boolean canEarlyTerminateOnDocId(Sort searchSort) {
+        final SortField[] fields1 = searchSort.getSort();
+        return SortField.FIELD_DOC.equals(fields1[0]);
+    }
+
+    private boolean canEarlyTerminateOnPrefix(Sort searchSort, Sort indexSort) {
+        if (indexSort != null) {
+            final SortField[] fields1 = searchSort.getSort();
+            final SortField[] fields2 = indexSort.getSort();
```

Comment:
please use more meaningful variable names instead of `fields1` and `fields2`

Reply:
ack